# Copyright 2025 Miljenko Å uflaj
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest

from amend.built_in.mappings.data_mappings import amend_mutable_data_mapping


class TestAmendMutableDataMapping(unittest.TestCase):
    def test_no_amendment_needed(
        self,
    ):
        try:
            amend_mutable_data_mapping(
                mutable_data_mapping={
                    "a": 1,
                    "b": 2,
                }
            )
        except Exception as e:
            self.assert_(
                False,
                f"Threw exception when no amendment was needed: {e}",
            )

    def test_type_mismatch(
        self,
    ):
        data_mapping = (
            ("a", 1),
            ("b", 2),
        )
        expected_result = {
            "a": 1,
            "b": 2,
        }

        try:
            result = amend_mutable_data_mapping(mutable_data_mapping=data_mapping)
        except Exception as e:
            self.assert_(
                False,
                f"Threw exception when amendment was possible: {e}",
            )
        self.assertEqual(
            result,
            expected_result,
        )

        self.assertRaises(
            TypeError,
            amend_mutable_data_mapping,
            mutable_data_mapping=data_mapping,
            type_mismatch_action="error",
        )

    def test_cast_error(
        self,
    ):
        self.assertRaises(
            TypeError,
            amend_mutable_data_mapping,
            mutable_data_mapping=None,
        )
        self.assertRaises(
            ValueError,
            amend_mutable_data_mapping,
            mutable_data_mapping=None,
            value_on_cast_error=1,
        )
        self.assertRaises(
            ValueError,
            amend_mutable_data_mapping,
            mutable_data_mapping=None,
            value_on_cast_error=(1, 2),
        )
        self.assertRaises(
            ValueError,
            amend_mutable_data_mapping,
            mutable_data_mapping=None,
            value_on_cast_error=(
                ("a", 1),
                2,
            ),
        )

        try:
            result = amend_mutable_data_mapping(
                mutable_data_mapping=None,
                value_on_cast_error=(
                    ("a", 1),
                    ("b", 2),
                ),
            )
        except Exception as e:
            self.assert_(
                False,
                f"Threw exception when value on cast error was valid: {e}",
            )

        self.assertEqual(
            result,
            {
                "a": 1,
                "b": 2,
            },
        )

    def test_case_with_length_violation(
        self,
    ):
        data_mapping = {
            "a": 1,
            "b": 2,
        }

        try:
            amend_mutable_data_mapping(
                mutable_data_mapping=data_mapping,
                length_violation_action="error",
            )
        except Exception as e:
            self.assert_(
                False,
                f"Threw exception for length violation without defined limits: {e}",
            )

        try:
            amend_mutable_data_mapping(
                mutable_data_mapping=data_mapping,
                minimum_length=0,
                length_violation_action="error",
            )
        except Exception as e:
            self.assert_(
                False,
                f"Threw exception for length violation with minimum 0 length: {e}",
            )

        try:
            amend_mutable_data_mapping(
                mutable_data_mapping=data_mapping,
                maximum_length=len(data_mapping),
                length_violation_action="error",
            )
        except Exception as e:
            self.assert_(
                False,
                "Threw exception for length violation with exact maximum length",
            )

        self.assertRaises(
            ValueError,
            amend_mutable_data_mapping,
            mutable_data_mapping=data_mapping,
            minimum_length=len(data_mapping) + 1,
            length_violation_action="error",
        )
        self.assertRaises(
            ValueError,
            amend_mutable_data_mapping,
            mutable_data_mapping=data_mapping,
            maximum_length=len(data_mapping) - 1,
            length_violation_action="error",
        )


if __name__ == "__main__":
    unittest.main()
